### PR TITLE
[Yul] Expression joiner.

### PR DESCRIPTION
**Pull request by chriseth**
Depends on https://github.com/ethereum/solidity/pull/5203

### CODE REVIEW

File: libjulia/optimiser/ExpressionUnbreaker.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+void ExpressionUnbreaker::handleArguments(vector<Expression>& _arguments)
+{
+	// We have to fill from left to right, but we can only
+	// fill if everything to the right is just an identifier
+	// or a literal.
+	// Also we only descend into function calls if everything
+	// on the right is an identifier or literal.
+
+	size_t numVarsOrLitsOnRight = 0;
```
Comment (chriseth) [inline libjulia/optimiser/ExpressionUnbreaker.cpp]: If someone has a suggestion about how to simplify this, please be my guest :)

File: libjulia/optimiser/ExpressionUnbreaker.cpp [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimiser component that undoes what the ExpressionUnbreaker did, i.e.
```
Comment (leonardoalt) [inline libjulia/optimiser/ExpressionUnbreaker.cpp]: What the `ExpressionBreaker` did

**Comment by leonardoalt** [comment]
`Unbreaker` sounds a bit weird to me. I understand that the goal is to keep it connected to the `ExpressionBreaker` as the inverse operation, just wanted to bring that up.
Would you consider other names, such as `ExpressionInliner`?

File: libjulia/optimiser/README.md [hunk 3]
```diff
@@ -56,6 +56,44 @@ depend on any side-effects.
...
+Note that this transformation does not change the order of opcodes or function calls.
+
+It is not applied to loop conditions, because the loop control flow does not allow
+this "outlining" of the inner expressions in all cases.
+
+The final program should be in a form such that with the exception of loop conditions,
+function calls can only appear in the right-hand side of a variable declaration,
+assignments or expression statements and all arguments have to be constants or variables.
+
+The benefits of this form is that it is much easier to re-order the sequence of opcodes
```
Comment (leonardoalt) [inline libjulia/optimiser/README.md]: `benefits ... are`

File: test/libjulia/ExpressionUnbreaker.cpp [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Unit tests for the expression breaker.
```
Comment (leonardoalt) [inline test/libjulia/ExpressionUnbreaker.cpp]: `unbreaker`

File: test/libjulia/ExpressionUnbreaker.cpp [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+		R"({
+			let a := mload(2)
+			let b := mload(a)
+			a := 4
+		})",
+		R"({
+			let a := mload(2)
+			let b := mload(a)
+			a := 4
+		})"
```
Comment (leonardoalt) [inline test/libjulia/ExpressionUnbreaker.cpp]: Why is there no inlining in this case?

File: test/libjulia/ExpressionUnbreaker.cpp [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+				x := add(a, 3)
+			}
+		})",
+		R"({
+			function f(a) -> x
+			{
+				a := mload(2)
+				x := add(a, 3)
+			}
+		})"
```
Comment (leonardoalt) [inline test/libjulia/ExpressionUnbreaker.cpp]: Why is there no inlining in this case?

File: libjulia/optimiser/ExpressionUnbreaker.cpp [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+	{
+		Identifier const& identifier = boost::get<Identifier>(_e);
+		if (isLatestStatementVarDeclOf(identifier) && m_references[identifier.name] == 1)
+		{
+			VariableDeclaration& varDecl = boost::get<VariableDeclaration>(*latestStatement());
+			assertThrow(varDecl.variables.size() == 1, OptimizerException, "");
+			assertThrow(varDecl.value, OptimizerException, "");
+
+			_e = std::move(*varDecl.value);
+			*latestStatement() = Block();
```
Comment (leonardoalt) [inline libjulia/optimiser/ExpressionUnbreaker.cpp]: Why is this necessary?

**Comment by chriseth** [comment]
Updated and renamed. I also made the constructor private and added a static function to prevent mis-use - not sure if it helps much, since the visitor functions have to be public.

**Comment by chriseth** [comment]
Will rebase onto #5203

**Review by leonardoalt — APPROVED** [review]
